### PR TITLE
Saving circuit.layout after transpiling

### LINKED ISSUE #2584: Need Consistent Layout Representations During Transpile
There are a few bug reports recently with how Layout is applied. The source of confusion is that different passes do slightly different thigns with the layout. 

After internal discussions with @itoko @kdk @1ucian0, here is the plan:

- [x] Clean up the ApplyLayout pass (#2050) and merge it. This pass should be dumb, and just consume a virtual circuit + Layout object, and output a register-less circuit (for now, a flat register). This involves updating the dag nodes as well as they contain register pointers. @itoko 

- [x] Change BasicSwap and StochasticSwap to consume physical circuits (register-less where each index literally corresponds to the physical qubit). @1ucian0 

- [x] Change LookaheadSwap to remove the logic for flattening register. @kdk 

- [x] Remove LegacySwap and replace by StochasticSwap (after some more testing). @ajavadia 

- [x] Update the other passes that currently consume a Layout (e.g. CXDirection, CXCheck, ..) to just assume that the layout has been applied, so just consult the (physical) circuit. @ajavadia 

- [ ] Add a circuit.layout property, so that the selected layout can be queried by the user later. This property will be populated by the layout selection pass (NoiseAdaptiveLayout, DenseLayout, TrivialLayout, SetLayout). So then, remove this from `pass_manager.property_set`. @1ucian0 

Fixing these should close  #1831, #2066, #2405, #2503, #2532.

### PR DESCRIPTION
**Pull request by 1ucian0**
This is an alternative to PR #2721 (and fixes #2584).

This change "saves" the layout into a circuit property after transpiling.

I would like to mentioning that, in the light of #2579, drawing the circuit during debugging will not include the layout using this approach.

### CODE REVIEW

File: test/python/transpiler/test_preset_passmanagers.py [hunk 1]
```diff
@@ -97,17 +107,24 @@ def test_initial_layout_2(self):
...
+                        6: ancilla[3], 7: ancilla[4], 8: ancilla[5], 9: qr[3], 10: ancilla[6],
+                        11: qr[2], 12: ancilla[7], 13: ancilla[8]}
         backend = FakeMelbourne()
 
         for optimization_level in range(4):
             qc_b = transpile(qc, backend, initial_layout=initial_layout,
                              optimization_level=optimization_level)
+            for wire in range(14):
+                with self.subTest(level=optimization_level, wire=wire):
+                    self.assertEqual(qc_b.layout[wire], final_layout[wire])
```
Comment (ajavadia) [inline test/python/transpiler/test_preset_passmanagers.py]: this doesn't really seem like a subtest. you just want to compare if two dictionaries are equal, not create a new test for each single wire!

File: test/python/transpiler/test_preset_passmanagers.py [hunk 2]
```diff
@@ -122,19 +139,31 @@ def test_initial_layout_3(self):
...
+                        19: ancilla[16]}
+
         backend = FakePoughkeepsie()
 
         for optimization_level in range(4):
             qc_b = transpile(qc, backend, initial_layout=initial_layout,
                              optimization_level=optimization_level)
+            for wire in range(14):
+                with self.subTest(level=optimization_level, wire=wire):
+                    self.assertEqual(qc_b.layout[wire], final_layout[wire])
```
Comment (ajavadia) [inline test/python/transpiler/test_preset_passmanagers.py]: same here.

File: test/python/transpiler/test_preset_passmanagers.py (hunk 1 above)
Comment (1ucian0) [reply test/python/transpiler/test_preset_passmanagers.py]: Indeed. I was combinant during the creation of the test. Removed in d9aabcc3

File: test/python/transpiler/test_preset_passmanagers.py (hunk 2 above)
Comment (1ucian0) [reply test/python/transpiler/test_preset_passmanagers.py]: Removed in d9aabcc3

**Pull request merged**